src/connectors/impls/gcs/writer.rs

```diff
@@ -0,0 +1,831 @@
+                            .as_str();
+
+                        // Range format: bytes=0-262143
+                        let range_end = &raw_range
+                            .get(
+                                raw_range
+                                    .find('-')
+                                    .ok_or(ErrorKind::GoogleCloudStorageError(
+                                        "Did not find a - in the Range header",
+                                    ))?
+                                    + 1..,
+                            )
+                            .ok_or(ErrorKind::GoogleCloudStorageError(
+                                "Unable to get the end of the Range",
+                            ))?;
+
+                        // NOTE: The data can only be thrown away to the point of the end of the Range header,
+                        // since google can persist less than we send in our request.
+                        //
+                        // see: https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+                        done_until.store(range_end.parse()?, Ordering::Release);
+                    } else {
+                        return Err("No range header".into());
+                    }
+                } else {
+                    return Err("no response from GCS".into());
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+struct GCSWriterSink {
+    #[cfg(test)]
+    client: Option<tests::MockHttpClient>,
+    #[cfg(not(test))]
+    client_tx: Option<Sender<HttpTaskCommand>>,
+    #[allow(unused)] // FIXME: use this in the http task
+    url: Url<HttpsDefaults>,
+    config: Config,
+    buffers: ChunkedBuffer,
+    current_name: Option<String>,
+    default_bucket: Option<Value<'static>>,
+    done_until: Arc<AtomicUsize>,
+}
+
+#[async_trait::async_trait]
+impl Sink for GCSWriterSink {
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        event: Event,
+        ctx: &SinkContext,
+        serializer: &mut EventSerializer,
+        _start: u64,
+    ) -> Result<SinkReply> {
+        self.buffers
+            .mark_done_until(self.done_until.load(Ordering::Acquire))?;
+
+        for (value, meta) in event.value_meta_iter() {
+            let meta = ctx.extract_meta(meta);
+
+            let name = meta
+                .get("name")
+                .ok_or(ErrorKind::GoogleCloudStorageError(
+                    "Metadata is missing the file name",
+                ))?
+                .as_str()
+                .ok_or(ErrorKind::GoogleCloudStorageError(
+                    "The file name in metadata is not a string",
+                ))?;
+
+            self.finish_upload_if_needed(name).await?;
+
+            self.start_upload_if_needed(meta, name).await?;
+
+            let serialized_data = serializer.serialize(value, event.ingest_ns)?;
+            for item in serialized_data {
+                self.buffers.write(&item);
```

Comment:
the tasks seem to be bucket/name based, allowing for multiple names to be uploaded, but there seems to be only one buffer, I feel like I'm missing something here. are the bucket/name tasks simply "future-proofing" and not relevant in `on_event`?

Reply:
The current way of working is that we're uploading all the events to a single buffer, until the name (and/or bucket) changes, and the `http_task` processes the events sequentially, so there's no need for multiple buffers.
I hope in the future we can support uploading multiple files in parallel, but this will require some careful analysis.

Reply:
makes perfect sense!